### ISSUE TITLE
fetchAuthSession triggering logout when offline

### ISSUE THREAD

**Issue by dehli** [aws-amplify/amplify-js#13999]
### Before opening, please confirm:

- [X] I have [searched for duplicate or closed issues](https://github.com/aws-amplify/amplify-js/issues?q=is%3Aissue+) and [discussions](https://github.com/aws-amplify/amplify-js/discussions).
- [X] I have read the guide for [submitting bug reports](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#bug-reports).
- [X] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.

### JavaScript Framework

React

### Amplify APIs

Authentication

### Amplify Version

v6

### Amplify Categories

auth

### Backend

CDK

### Environment information

<details>

```
# Put output below this line


```

</details>

### Describe the bug

If you call `Auth.fetchAuthSession({ forceRefresh: true })` while offline, amplify will trigger the logout flow.

### Expected behavior

I'd expect an error, but I wouldn't expect to be logged out.

### Reproduction steps

1. Install latest version of AWS amplify with auth.
2. Sign in.
3. Go offline.
4. Call `Auth.fetchAuthSession({ forceRefresh: true })` (I did this by adding the function to the window).

### Code Snippet

```javascript
// Put your code below this line.
window.TRIGGER_ERROR = () => Auth.fetchAuthSession({ forceRefresh: true });
```

### Log output

```js
// Put your logs below this line
{"channel":"auth","payload":{"event":"tokenRefresh_failure","data":{"error":{"name":"NetworkError","underlyingError":{}}}},"source":"Auth","patternInfo":[]}
```

### Additional information and screenshots

Related to https://github.com/aws-amplify/amplify-js/issues/13596

**Comment by dehli** [comment]
Closing this as I'm investigating the issue. Will reopen when I have more information.

**Comment by didemkkaslan** [comment]
Having the same issue,  I see tokenRefresh_failure event being fired and tokens are removed from cookie storage

**Comment by ashika112** [comment]
@dehli Thanks for taking a look. I can confirm i see the same behavior as well. I will look closer into this today and report back.

**Comment by ashika112** [comment]
@dehli @didemkkaslan Can you confirm which version are you both on? There was a retry fix just recently, while i see the issue on 6.6.4 (I used this version since that was what @didemkkaslan issue mentioned and on another issue as well) when i try the most recent version 6.8.0 i dont see the logout behavior anymore. Can one of you confirm ?

**Comment by dehli** [comment]
Hey @ashika112, thanks for looking into it. I'm on `6.8.0` currently and seeing this behavior. I just removed my `node_modules` and reinstalled them to confirm that I was on that version. I was still able to trigger logout by calling `fetchAuthSession({ forceRefresh: true });` while offline.

**Comment by ashika112** [comment]
@dehli so this what i see, i do get network error but after that when network comes back on i am able to get credentials and token for the customer and not logged out.
![Screenshot 2024-11-11 at 11 19 16 AM](https://github.com/user-attachments/assets/938f2110-a86b-4990-99d6-eed81de80de1)

**Comment by ashika112** [comment]
@dehli do you have any custom storage implementation? Would it be possible to put out a minimal code reproducing and share me the repo?

**Comment by dehli** [comment]
No custom storage. I'll work on putting together a minimal reproducing example though 👍

**Comment by ashika112** [comment]
@dehli , I was also testing in next js since some other issues was there. I will also try in react now at my end (which i notice is what you use at your end).

**Comment by didemkkaslan** [comment]
Updated the versions to 

 "aws-amplify": "^6.8.0",
 "@aws-amplify/adapter-nextjs": "^1.2.26",
 
 and while offline I've used the same window.TRIGGER_ERROR method @dehli  uses  (fetchAuthSession({ tokenRefresh: force }) below is the error I'm getting and tokens are deleted from my custom cookie storage. 

<img width="1468" alt="Screenshot 2024-11-11 at 22 42 41" src="https://github.com/user-attachments/assets/379f9393-91bd-4a19-bcbe-86a1b7e548e5">


Custom cookie storage: 
```
const cookieOptions: OptionsType =
  process.env.NEXT_PUBLIC_ENV === 'msteams'
    ? {
        domain: 'tab.app.spiky.ai' as string,
        sameSite: 'none' as 'lax' | 'strict' | 'none',
        secure: true,
      }
    : {};

const keyValueStorage = createKeyValueStorageFromCookieStorageAdapter({
  get(name) {
    const value = getCookie(name, cookieOptions);
    return { name, value };
  },
  getAll() {
    const cookies = getCookies(cookieOptions);
    return Object.keys(cookies).map((name) => ({ name, value: cookies[name] }));
  },
  set(name, value) {
    const cookieAlreadyExists = getCookie(name, cookieOptions);

    if (cookieAlreadyExists) {
      console.log(
        'cookieAlreadyExists so Im deleting it and setting new one',
        cookieAlreadyExists,
      );
      deleteCookie(name, cookieOptions);
    }
    setCookie(name, value, cookieOptions);
  },
  delete(name) {
    deleteCookie(name, cookieOptions);
  },
});
```

**Comment by ashika112** [comment]
@dehli Im able to reproduce this consistently in React and hence i have marked this as bug. We will keep this ticket posted for updates.

**Comment by dehli** [comment]
Awesome! Thanks so much 🎉

**Comment by didemkkaslan** [comment]
I've tested both versions

**With    aws-amplify: "6.6.4"**
When I go offline and triggered the method @dehli  provides `window.TRIGGER_ERROR = () => Auth.fetchAuthSession({ forceRefresh: true });
`

Got the below error and tokens are deleted from my custom cookie storage
<img width="1470" alt="Screenshot 2024-11-11 at 23 03 18" src="https://github.com/user-attachments/assets/c09ab5b1-cb49-41db-aa9b-2e0120eb627f">
<img width="1470" alt="Screenshot 2024-11-11 at 23 03 39" src="https://github.com/user-attachments/assets/3f74acdd-b362-41d8-bd79-065fc2c3230f">

**With    aws-amplify: "6.6.4"**

Tokens are not deleted from my custom cookie storage, I can still access session idtoken data

<img width="1470" alt="Screenshot 2024-11-11 at 23 09 08" src="https://github.com/user-attachments/assets/b7271b4e-1625-4279-b5c6-b64c9d2d63f8">
<img width="1467" alt="Screenshot 2024-11-11 at 23 09 34" src="https://github.com/user-attachments/assets/0914c88b-e7f5-403c-b418-429f6438a069">


I have one question tho: I shouldn't signout the user when a tokenrefresh_failure event fires right?

**Comment by ashika112** [comment]
@didemkkaslan Ideally you want log customer out when there is error except for Network error.

**Comment by ashika112** [comment]
@dehli Do you use `Authenticator` by any chance? taking another close look. I see this behavior only when using Authenticator component not when using the Auth APIs directly? If you do not use Authenticator, then I am going to need the reproducible code sample.

**Comment by dehli** [comment]
@ashika112  Yep, I am using the `Authenticator` component (although I am referencing `Auth.fetchAuthSession` directly). No need to apologize, happy to help how I can.

**Comment by ashika112** [comment]
@dehli I think this is scoped to Authenticator. Is it possible for you to verify the behavior without Authenticator? On my side, im going to reach out to `amplify-ui` repo.

**Comment by dehli** [comment]
It will take a bit to pull out the Authenticator with the way our project is setup. I did look through the `amplify-js` library for a bit this morning a bit and couldn't find anything that would cause the signout behavior so it being in a separate repo makes a lot of sense. 

From my investigation this morning, it seemed to be related to the `tokenRefresh_failure` event that is sent over `Hub`. When searching over there for it, you can see a handful of references so I think it's safe to close this issue. https://github.com/search?q=repo%3Aaws-amplify%2Famplify-ui+tokenRefresh_failure&type=code

**Comment by ashika112** [comment]
@dehli I am working internally with the team right now, if this needs to be in their repo, i will re-route this ticket to them so they have context.

**Comment by dehli** [comment]
Sounds good! This seems to be the offending line:

https://github.com/aws-amplify/amplify-ui/blob/2e3a051d9c2d48d5be22d2e83095cbd060ce2b40/packages/ui/src/helpers/authenticator/defaultAuthHubHandler.ts#L32-L36
